src/EditorFeatures/Core/Wrapping/AbstractParameterWrappingCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Editor.Wrapping
+{
+    internal abstract partial class AbstractWrappingCodeRefactoringProvider<
+        TListSyntax,
+        TListItemSyntax>
+        : CodeRefactoringProvider
+        where TListSyntax : SyntaxNode
+        where TListItemSyntax : SyntaxNode
+    {
+        private static ImmutableArray<string> s_mruTitles = ImmutableArray<string>.Empty;
+
+        protected abstract string ListName { get; }
+        protected abstract string ItemNamePlural { get; }
+        protected abstract string ItemNameSingular { get; }
+
+        protected abstract TListSyntax GetApplicableList(SyntaxNode node);
+        protected abstract SeparatedSyntaxList<TListItemSyntax> GetListItems(TListSyntax listSyntax);
+        protected abstract bool PositionIsApplicable(
+            SyntaxNode root, int position, SyntaxNode declaration, TListSyntax listSyntax);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var span = context.Span;
+            if (!span.IsEmpty)
+            {
+                return;
+            }
+
+            var position = span.Start;
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(position);
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var generator = document.GetLanguageService<SyntaxGenerator>();
+
+            var declaration = token.Parent.Ancestors().FirstOrDefault(n => GetApplicableList(n) != null);
+            if (declaration == null)
+            {
+                return;
+            }
+
+            var listSyntax = GetApplicableList(declaration);
+            // Make sure we don't have any syntax errors here.  Don't want to format if we don't
+            // really understand what's going on.
```

Comment:
even if the syntax error is inside an anonymous function boundary?

Reply:
i'm ok with that.